it/src/main/java/org/corfudb/universe/universe/vm/VmUniverse.java

```diff
@@ -0,0 +1,319 @@
+     */
+    private VirtualMachine deployVm(String vmName) throws NodeException {
+        log.info("Deploy VM: {}", vmName);
+
+        VirtualMachine vm;
+        try {
+            // Connect to vSphere server using VIJAVA
+            VmUniverseParams params = universeParams;
+            ServiceInstance si = new ServiceInstance(
+                    new URL(params.getVSphereUrl()),
+                    params.getVSphereUsername(),
+                    params.getVSpherePassword(),
+                    true
+            );
+
+            InventoryNavigator inventoryNavigator = new InventoryNavigator(si.getRootFolder());
+
+            // First check if a VM with this name already exists or not
+            vm = (VirtualMachine) inventoryNavigator.searchManagedEntity(
+                    ManagedEntityType.VIRTUAL_MACHINE.typeName, vmName);
+            if (vm != null) {
+                // If the VM already exists, ensure the power state is 'on' before return the VM
+                if (vm.getSummary().runtime.powerState == VirtualMachinePowerState.poweredOff) {
+                    log.info(vmName + " already exists, but found in 'poweredOff' state, powering it on...");
+                    Task task = vm.powerOnVM_Task((HostSystem) inventoryNavigator.searchManagedEntity(
+                            "HostSystem", universeParams.getVSphereHost()));
+                    task.waitForTask();
+                }
+            } else {
+                // Find the template machine in the inventory
+                VirtualMachine vmTemplate = (VirtualMachine) inventoryNavigator.searchManagedEntity(
+                        ManagedEntityType.VIRTUAL_MACHINE.typeName,
+                        params.getTemplateVMName()
+                );
+
+                log.info("Deploying the VM {} via vSphere {}...", vmName, params.getVSphereUrl());
+
+                // Create customization for cloning process
+                VirtualMachineCloneSpec cloneSpec = createLinuxCustomization(vmName);
+                try {
+                    // Do the cloning - providing the clone specification
+                    Task cloneTask = vmTemplate.cloneVM_Task((Folder) vmTemplate.getParent(), vmName, cloneSpec);
+                    cloneTask.waitForTask();
+                } catch (RemoteException | InterruptedException e) {
+                    throw new UniverseException(String.format("Deploy VM %s failed due to ", vmName), e);
+                }
+                // After the clone task completes, get the VM from the inventory
+                vm = (VirtualMachine) inventoryNavigator.searchManagedEntity(
+                        ManagedEntityType.VIRTUAL_MACHINE.typeName, vmName);
+            }
+            // Ensure we get the VM's IP address before we return its instance
+            String cloneVmIpAddress = null;
+            log.info("Getting IP address for {} from DHCP...please wait...", vmName);
+            while (cloneVmIpAddress == null) {
```

Comment:
Add a debug log in the loop to make sure we detect if we are in an infinite loop.

Reply:
Fixed